File: 2021/day07/Program.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;

public class Program
{
    static void Main(string[] args)
    {
        if (args.Length > 0)
        {
            string fileName = args[0];
            if (File.Exists(fileName))
            {
                Stopwatch stopWatch = new Stopwatch();
                stopWatch.Start();
                string[] lines = File.ReadAllLines(fileName);
                stopWatch.Stop();
                Console.WriteLine($"File read ({stopWatch.Elapsed.TotalMilliseconds} ms)");

                Stopwatch part1Timer = new Stopwatch();
                part1Timer.Start();
                long part1 = SolvePart1(lines);
                part1Timer.Stop();
                Console.WriteLine($"Part 1: {part1} ({part1Timer.Elapsed.TotalMilliseconds} ms)");

                Stopwatch part2Timer = new Stopwatch();
                part2Timer.Start();
                long part2 = SolvePart2(lines);
                part2Timer.Stop();
                Console.WriteLine($"Part 2: {part2} ({part2Timer.Elapsed.TotalMilliseconds} ms)");
            }
            else
            {
                throw new ArgumentException("Invalid file name provided. Please provide a valid file name.");
            }
        }
        else
        {
            throw new ArgumentException("Input data file name not provided. Please provide the file name as an argument: dotnet run <file-name>");
        }
    }

    static long SolvePart1(string[] lines)
    {
        List<long> numbers = ParseInput(lines);

        long min = long.MaxValue;
        long max = long.MinValue;
        foreach (long number in numbers)
        {
            if (number > max) max = number;
            if (number < min) min = number;
        }

        long minFuel = long.MaxValue;
        for (long position = min; position <= max; position++)
        {
            long total = 0;
            foreach (long number in numbers)
            {
                long fuelUsed = Math.Abs(number - position);
                total += fuelUsed;

                if (total > minFuel) break;
            }
            if (total < minFuel) minFuel = total;
        }

        return minFuel;
    }

    static long SolvePart2(string[] lines)
    {
        List<long> numbers = ParseInput(lines);

        long min = long.MaxValue;
        long max = long.MinValue;
        foreach (long number in numbers)
        {
            if (number > max) max = number;
            if (number < min) min = number;
        }

        long minFuel = long.MaxValue;
        for (long position = min; position <= max; position++)
        {
            long total = 0;
            foreach (long number in numbers)
            {
                long fuelUsed = Math.Abs(number - position);
                total += fuelUsed * (fuelUsed + 1) / 2;

                if (total > minFuel) break;
            }
            if (total < minFuel) minFuel = total;
        }

        return minFuel;
    }

    static List<long> ParseInput(string[] lines)
    {
        List<long> numbers = new();

        foreach (string line in lines)
        {
            foreach (string item in line.Split(','))
            {
                numbers.Add(long.Parse(item));
            }
        }

        return numbers;
    }
}

File: 2021/day07/program.go
package day07

import (
	"bytes"
	"fmt"
	"math"
	"time"

	"github.com/jasonncleveland/adventofcode/2021/utils"
)

func Run(fileName string) {
	var start time.Time = time.Now()
	lines := utils.ReadFileLines(fileName)
	fmt.Printf("File read: %s\n", time.Since(start))

	start = time.Now()
	part1 := Part1(lines)
	fmt.Printf("Part 1: %d (%s)\n", part1, time.Since(start))

	start = time.Now()
	part2 := Part2(lines)
	fmt.Printf("Part 2: %d (%s)\n", part2, time.Since(start))
}

func Part1(lines [][]byte) int64 {
	numbers := ParseInput(lines)[0]

	var max int64 = math.MinInt64
	var min int64 = math.MaxInt64

	for _, number := range numbers {
		if number > max {
			max = number
		}
		if number < min {
			min = number
		}
	}

	var minFuel int64 = math.MaxInt64
	for step := range max {
		total := int64(0)
		for _, crab := range numbers {
			fuel := crab - step
			if fuel < 0 {
				// Ensure value is positive
				fuel *= -1
			}
			total += fuel

			// Exit early if we've already surpassed the min fuel found
			if total > minFuel {
				break
			}
		}
		if total < minFuel {
			minFuel = total
		}
	}

	return minFuel
}

func Part2(lines [][]byte) int64 {
	numbers := ParseInput(lines)[0]

	var max int64 = math.MinInt64
	var min int64 = math.MaxInt64

	for _, number := range numbers {
		if number > max {
			max = number
		}
		if number < min {
			min = number
		}
	}

	var minFuel int64 = math.MaxInt64
	for step := range max {
		total := int64(0)
		for _, crab := range numbers {
			fuel := crab - step
			if fuel < 0 {
				// Ensure value is positive
				fuel *= -1
			}
			fuel = fuel * (fuel + 1) / 2
			total += fuel

			// Exit early if we've already surpassed the min fuel found
			if total > minFuel {
				break
			}
		}
		if total < minFuel {
			minFuel = total
		}
	}

	return minFuel
}

func ParseInput(lines [][]byte) [][]int64 {
	var data [][]int64

	for _, line := range lines {
		var numbers []int64
		for _, number := range bytes.Split(line, []byte(",")) {
			numbers = append(numbers, ParseNumber(number))
		}
		data = append(data, numbers)
	}

	return data
}
